logging/logger.go

```diff
@@ -0,0 +1,293 @@
+package logging
+
+import (
+	"sync"
+	"testing"
+	"time"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"go.uber.org/zap/zaptest"
+	"go.uber.org/zap/zaptest/observer"
+)
+
+// Logger interface for logging to.
+type Logger interface {
+	ZapCompatibleLogger
+
+	AsZap() *zap.SugaredLogger
+}
+
+// ZapCompatibleLogger is a backwards compatibility layer for existing usages of the RDK as a
+// library for Go application code or modules. Public (to the library) methods that take a logger as
+// input should accept this type and upconvert to a Logger via a call to `FromZapCompatible`.
+type ZapCompatibleLogger interface {
+	Desugar() *zap.Logger
+	Level() zapcore.Level
+	Named(name string) *zap.SugaredLogger
+	Sync() error
+	With(args ...interface{}) *zap.SugaredLogger
+	WithOptions(opts ...zap.Option) *zap.SugaredLogger
+
+	Debug(args ...interface{})
+	Debugf(template string, args ...interface{})
+	Debugln(args ...interface{})
+	Debugw(msg string, keysAndValues ...interface{})
+
+	Info(args ...interface{})
+	Infof(template string, args ...interface{})
+	Infoln(args ...interface{})
+	Infow(msg string, keysAndValues ...interface{})
+
+	Warn(args ...interface{})
+	Warnf(template string, args ...interface{})
+	Warnln(args ...interface{})
+	Warnw(msg string, keysAndValues ...interface{})
+
+	Error(args ...interface{})
+	Errorf(template string, args ...interface{})
+	Errorln(args ...interface{})
+	Errorw(msg string, keysAndValues ...interface{})
+
+	Fatal(args ...interface{})
+	Fatalf(template string, args ...interface{})
+	Fatalln(args ...interface{})
+	Fatalw(msg string, keysAndValues ...interface{})
+
+	Panic(args ...interface{})
+	Panicf(template string, args ...interface{})
+	Panicln(args ...interface{})
+	Panicw(msg string, keysAndValues ...interface{})
+
+	DPanic(args ...interface{})
+	DPanicf(template string, args ...interface{})
+	DPanicln(args ...interface{})
+	DPanicw(msg string, keysAndValues ...interface{})
+}
+
+// ZLogger type for logging to. Wraps a zap logger and adds the `AsZap` method to satisfy the
+// `Logger` interface.
+type zLogger struct {
+	*zap.SugaredLogger
+}
+
+// FromZapCompatible upconverts a ZapCompatibleLogger to a logging.Logger. If the argument already
+// satisfies logging.Logger, no changes will be made. A nil input returns a nil logger. An input of
+// unknown type will create a new logger that's not associated with the input.
+func FromZapCompatible(logger ZapCompatibleLogger) Logger {
+	if logger == nil {
+		return nil
+	}
+
+	switch l := logger.(type) {
+	case *zap.SugaredLogger:
+		// golog.Logger is a type alias for *zap.SugaredLogger and is captured by this.
+		return zLogger{l}
+	case Logger:
+		return l
+	default:
+		logger.Warnf("Unknown logger type, creating a new Viam Logger. Unknown type: %T", logger)
+		return newDefaultLogger()
+	}
+}
+
+// AsZap converts the logger to a zap logger.
+func (logger zLogger) AsZap() *zap.SugaredLogger {
```

Comment:
I'm sure I'm missing the whole point of this interface-to-maintain-backward-compatability-system by asking this question, but _why_ is `AsZap()` needed? I see you using it everywhere to extract the sugared logger from the `zLogger`, but that sugared logger is a promoted field; shouldn't all method calls to the `zLogger` (besides `AsZap`) route directly to the sugared logger and shouldn't `zLogger` implement both `Logger` and `ZapCompatibleLogger`? I think I understand the "up" conversion from a `ZapCompatibleLogger` to a `Logger`, but I don't understand the point of this "down" conversion.

Reply:
I did see this:

> Additionally to needing to accept zap loggers in some places, there are some places where we need to "downconvert" an RDK logging.Logger back to a zap logger. This is when calling out to other (viam) libraries, such as [gostream](https://github.com/viamrobotics/rdk/blob/b419881e6383bb98d20a61180075681bb6f67f7b/components/audioinput/microphone/microphone.go#L116) that accept zap/golog loggers as arguments.

Which I think answers my question about _why_ `AsZap` is needed. But now I'm confused why we can't call `AsZap()` before those those calls to external libraries that only accept concrete zap types instead of having `client.New` only accept the concrete zap type.

Reply:
I think `Client.New` accepting a sugared logger was some typo/mistake that you stumbled across. A pretty bad mistake where I suspect seeing that would greatly confuse you. Now that that's been corrected, is there still a misunderstanding?

The only usages for `AsZap` should be calling into external libraries that only accept the concrete `*zap.SugaredLogger`/`golog.Logger`. E.g: [gostreams](https://github.com/viamrobotics/rdk/blob/39e81fc74f3b302228ea1093cb389f7fa80184d6/components/audioinput/microphone/microphone.go#L116) or [goutils](https://github.com/viamrobotics/rdk/blob/39e81fc74f3b302228ea1093cb389f7fa80184d6/config/reader.go#L605).

Happy to go over this again now that you've taken a look. But want to make sure we're in agreement with what the code actually is!

Reply:
> Now that that's been corrected, is there still a misunderstanding?

Nope! I originally didn't understand why `AsZap` was needed at all, but I do now. And, it makes much more sense that `client.New` takes a `ZapCompatibleLogger` for backwards-compatible usages of the SDK.